json_logic/__init__.py

```diff
@@ -1,201 +1,457 @@
-from __future__ import unicode_literals
+"""
+This is a Python implementation of the JsonLogic JS library:
+https://github.com/jwadhams/json-logic-js
+"""
 
-from six.moves import reduce
-import logging
 
-logger = logging.getLogger(__name__)
-
-# Python 2 fallback
+# Python 2 fallbacks
+from __future__ import division, unicode_literals
 try:
     str = unicode
 except NameError:
     pass
+from six.moves import reduce
+from six import integer_types
+numeric_types = integer_types + (float,)
 
+import logging
+import warnings
 
-def if_(*args):
-    """Implements the 'if' operator with support for multiple elseif-s."""
-    for i in range(0, len(args) - 1, 2):
-        if args[i]:
-            return args[i + 1]
-    if len(args) % 2:
-        return args[-1]
-    else:
-        return None
+
+# Helper functions and variables
+
+# Sentinel value to indicate an optional argument
+# that can take various values including None
+_no_argument = object()
 
 
-def soft_equals(a, b):
-    """Implements the '==' operator, which does type JS-style coercion."""
+def _is_numeric(arg):
+    """Check if argument is of a numeric type: float, int or long."""
+    return type(arg) in numeric_types
+
+
+def _to_numeric(arg):
+    """
+    Convert a string or other value  into float, integer or long.
+    Convert float value to integer if appropriate.
+    """
+    if isinstance(arg, str) and '.' in arg:
+        arg = float(arg)
+    if isinstance(arg, float):
+        return int(arg) if arg.is_integer() else arg
+    return int(arg)
+
+
+# Common operations
+
+def _equal_to(a, b):
+    """Test for non-strict equality ('==') with JS-style type coercion."""
     if isinstance(a, str) or isinstance(b, str):
         return str(a) == str(b)
     if isinstance(a, bool) or isinstance(b, bool):
         return bool(a) is bool(b)
     return a == b
 
 
-def hard_equals(a, b):
-    """Implements the '===' operator."""
-    if type(a) != type(b):
-        return False
-    return a == b
+def _strict_equal_to(a, b):
+    """
+    Check for strict equality ('===') including type equality.
+    N.B.: Core JsonLogic does not differentiate between different numeric
+    types and considers them strictly equal to each other.
+    E.g.: {"===": [1, 1.0]}
+    """
+    if type(a) is type(b):
+        return a == b
+    if _is_numeric(a) and _is_numeric(b):
+        return _to_numeric(a) == _to_numeric(b)
+    return False
 
 
-def less(a, b, *args):
-    """Implements the '<' operator with JS-style type coercion."""
+def _not_equal_to(a, b):
+    """Check for non-strict inequality ('==') with JS-style type coercion."""
+    return not _equal_to(a, b)
+
+
+def _not_strict_equal_to(a, b):
+    """Check for strict inequality ('!==') including type inequality."""
+    return not _strict_equal_to(a, b)
+
+
+def _less_than(a, b, c=_no_argument):
+    """
+    Check that A is less then B (A < B) or
+    that B is exclusively between A and C (A < B < C).
```

Comment:
As I was looking at this, I realized the existing/unchanged code is not consistent with the javascript implementation. It does not check if values are numeric, it just checks if a < b and b < c if c exists. So it will handle regular strings as well as numeric values.
I recognize that you didn't make it work this way but your statement was making operators behave like current jsonlogic.